File: apps/HealthApp/android/app/src/main/java/com/myhealth/app/vision/LabReportOcr.kt
package com.myhealth.app.vision

import android.graphics.Bitmap
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * On-device lab report OCR. Mirrors iOS `LabReportOCR.swift` — same
 * schema, so iOS + Android produce identical fields. Snap a printed
 * Quest / LabCorp / Kaiser report → A1C, fasting glucose, BP, lipid
 * panel, BMI, weight.
 */
@Singleton
class LabReportOcr @Inject constructor() {
    private val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

    data class Result(
        val a1c: Double? = null,
        val fastingGlucose: Double? = null,
        val bpSystolic: Int? = null,
        val bpDiastolic: Int? = null,
        val cholesterolTotal: Double? = null,
        val ldl: Double? = null,
        val hdl: Double? = null,
        val triglycerides: Double? = null,
        val bmi: Double? = null,
        val weightKg: Double? = null,
        val rawText: String = "",
    )

    suspend fun read(bitmap: Bitmap, rotationDegrees: Int = 0): Result =
        suspendCancellableCoroutine { cont ->
            val image = InputImage.fromBitmap(bitmap, rotationDegrees)
            recognizer.process(image)
                .addOnSuccessListener { v ->
                    cont.resume(extract(v.text))
                }
                .addOnFailureListener { e -> cont.resumeWithException(e) }
        }

    suspend fun extract(text: String): Result {
        val ext = StructuredExtractorRegistry.shared
        val map = ext.extract(text, SCHEMA)
        val (sys, dia) = parseBp(map["bp"])
        return Result(
            a1c = map["a1c"]?.toDoubleOrNull(),
            fastingGlucose = map["fastingGlucose"]?.toDoubleOrNull(),
            bpSystolic = sys,
            bpDiastolic = dia,
            cholesterolTotal = map["cholesterol"]?.toDoubleOrNull(),
            ldl = map["ldl"]?.toDoubleOrNull(),
            hdl = map["hdl"]?.toDoubleOrNull(),
            triglycerides = map["triglycerides"]?.toDoubleOrNull(),
            bmi = map["bmi"]?.toDoubleOrNull(),
            weightKg = map["weight"]?.toDoubleOrNull(),
            rawText = text,
        )
    }

    private fun parseBp(s: String?): Pair<Int?, Int?> {
        s ?: return null to null
        val parts = s.split("/")
        if (parts.size != 2) return null to null
        return parts[0].trim().toIntOrNull() to parts[1].trim().toIntOrNull()
    }

    companion object {
        val SCHEMA = listOf(
            Field("a1c", listOf("a1c", "hba1c", "hemoglobin a1c"), """\d+\.\d+"""),
            Field("fastingGlucose",
                listOf("fasting glucose", "fasting blood sugar", "fbg", "glucose"),
                """\d+(\.\d+)?"""),
            Field("bp", listOf("bp", "blood pressure"), """\d{2,3}\s*/\s*\d{2,3}"""),
            Field("cholesterol",
                listOf("total cholesterol", "cholesterol total", "tc:"),
                """\d+(\.\d+)?"""),
            Field("ldl", listOf("ldl"), """\d+(\.\d+)?"""),
            Field("hdl", listOf("hdl"), """\d+(\.\d+)?"""),
            Field("triglycerides", listOf("triglycerides", "tg:"), """\d+(\.\d+)?"""),
            Field("bmi", listOf("bmi"), """\d+(\.\d+)?"""),
            Field("weight", listOf("weight"), """\d+(\.\d+)?"""),
        )
    }
}

File: apps/HealthApp/shared/FitFusionCore/Sources/FitFusionCore/Intelligence/LabReportOCR.swift
import Foundation
#if canImport(UIKit) && canImport(Vision)
import UIKit
import Vision

/// On-device lab report OCR. Snap a printed lab summary (Quest, LabCorp,
/// Kaiser) → extract the values Care+ surfaces in the per-condition
/// CarePlanCard: A1C, fasting glucose, BP systolic/diastolic, total
/// cholesterol, LDL, HDL, triglycerides, BMI, weight.
///
/// Same Vision pipeline as `NutritionLabelOCR` and `InsuranceCardOCR`;
/// new shape only differs in the schema fed to `StructuredExtractor`.
@MainActor
public final class LabReportOCR {

    public static let shared = LabReportOCR()
    private init() {}

    public struct Result: Sendable {
        public var a1c: Double?
        public var fastingGlucose: Double?
        public var bpSystolic: Int?
        public var bpDiastolic: Int?
        public var cholesterolTotal: Double?
        public var ldl: Double?
        public var hdl: Double?
        public var triglycerides: Double?
        public var bmi: Double?
        public var weightKg: Double?
        public var rawText: String

        public init(rawText: String = "") { self.rawText = rawText }
    }

    public func parse(image: UIImage) async -> Result {
        guard let cg = image.cgImage else { return .init() }
        let raw = await ocr(cg)
        return await Self.extract(from: raw)
    }

    /// Pure-function extraction so it's unit-testable from text fixtures.
    public static func extract(from text: String) async -> Result {
        let extractor = StructuredExtractorRegistry.shared
        let map = await extractor.extract(text: text, schema: schema)
        var r = Result(rawText: text)
        r.a1c = double(map["a1c"] ?? nil)
        r.fastingGlucose = double(map["fastingGlucose"] ?? nil)
        r.cholesterolTotal = double(map["cholesterol"] ?? nil)
        r.ldl = double(map["ldl"] ?? nil)
        r.hdl = double(map["hdl"] ?? nil)
        r.triglycerides = double(map["triglycerides"] ?? nil)
        r.bmi = double(map["bmi"] ?? nil)
        r.weightKg = double(map["weight"] ?? nil)
        if let bp = map["bp"] ?? nil {
            // Expect "138/88" style.
            let parts = bp.split(separator: "/")
            if parts.count == 2 {
                r.bpSystolic = Int(parts[0].trimmingCharacters(in: .whitespaces))
                r.bpDiastolic = Int(parts[1].trimmingCharacters(in: .whitespaces))
            }
        }
        return r
    }

    private static let schema: [Field] = [
        Field(name: "a1c",
              keywords: ["a1c", "hba1c", "hemoglobin a1c"],
              valuePattern: #"\d+\.\d+"#),
        Field(name: "fastingGlucose",
              keywords: ["fasting glucose", "fasting blood sugar", "fbg", "glucose"],
              valuePattern: #"\d+(\.\d+)?"#),
        Field(name: "bp",
              keywords: ["bp", "blood pressure"],
              valuePattern: #"\d{2,3}\s*/\s*\d{2,3}"#),
        Field(name: "cholesterol",
              keywords: ["total cholesterol", "cholesterol total", "tc:"],
              valuePattern: #"\d+(\.\d+)?"#),
        Field(name: "ldl",
              keywords: ["ldl"],
              valuePattern: #"\d+(\.\d+)?"#),
        Field(name: "hdl",
              keywords: ["hdl"],
              valuePattern: #"\d+(\.\d+)?"#),
        Field(name: "triglycerides",
              keywords: ["triglycerides", "tg:"],
              valuePattern: #"\d+(\.\d+)?"#),
        Field(name: "bmi",
              keywords: ["bmi"],
              valuePattern: #"\d+(\.\d+)?"#),
        Field(name: "weight",
              keywords: ["weight"],
              valuePattern: #"\d+(\.\d+)?"#),
    ]

    private func ocr(_ cg: CGImage) async -> String {
        await withCheckedContinuation { cont in
            let req = VNRecognizeTextRequest { req, _ in
                let lines = ((req.results as? [VNRecognizedTextObservation]) ?? [])
                    .compactMap { $0.topCandidates(1).first?.string }
                cont.resume(returning: lines.joined(separator: "\n"))
            }
            req.recognitionLevel = .accurate
            req.usesLanguageCorrection = true
            let h = VNImageRequestHandler(cgImage: cg, orientation: .up, options: [:])
            DispatchQueue.global(qos: .userInitiated).async {
                do { try h.perform([req]) } catch { cont.resume(returning: "") }
            }
        }
    }

    private static func double(_ s: String?) -> Double? {
        guard let s = s?.replacingOccurrences(of: ",", with: ".") else { return nil }
        return Double(s)
    }
}
#endif
